### DISCUSSION TITLE
Coupling Tensor Mechanics with THM Heat Structure

### DISCUSSION THREAD

**Discussion by jcstonehillAMA** [idaholab/moose#21724] [Q&A Modules: Solid mechanics] [answered]
Hello friends,

Hopefully this is just a silly mistake by me. I'm trying to use the THM Component "HeatStructureFromFile3D" to import an exodus file to create a mesh for use with some other THM components.

In addition, I'd like to use the Tensor Mechanics module to calculate some the stress/strain from both thermal expansion and an applied load.

Unfortunately, I can not get past the first step of simply getting the Tensor Mechanics module to recognize the mesh. My suspicion is that it's looking for a mesh that technically hasn't been created yet since it is generated by the HeatStructureFromFile3D component after this check. But, I'd love to hear if anyone has any feedback.

Code Snippet:
```
[Modules/TensorMechanics/Master]
	[all]
		add_variables = true
	[]
[]


[Components]

	[blk]
		type = HeatStructureFromFile3D
		file = 'GenerateFuelElement_in.e'
		position = '0 0 0'
		initial_T = 372
	[]
[]
```

EDIT: Forgot to include the error: 

```
*** ERROR ***
No subdomains found
```

**Comment by GiudGiud** [comment]
Hello

I dont think THM meshes from components are currently compatible with other modules. @joshuahansel for confirmation

Guillaume

**Reply by joshuahansel** [reply]
I'm not entirely sure, but I think that this is probably the case. THM was designed to couple with other physics via MultiApps or have the other apps write their own `Component`s. I can at least say that we have no examples of doing what is attempted here.

**Reply by LagrangeW** [reply]
Hello, Is there any example inputs on coupling THM using MultiApps? Ask this because THM module seems different with others that no [Variables] or [Kernels] is needed in a THM input.

**Reply by GiudGiud** [reply]
Hello

There's this example on the virtual test bed
https://github.com/idaholab/virtual_test_bed/tree/devel/htgr/assembly
thm.i is the THM input

Guillaume

**Reply by GiudGiud** [reply]
this is the doc page for it 
https://mooseframework.inl.gov/virtual_test_bed/htgr/assembly/index.html

**Reply by LagrangeW** [reply]
Got it! Thank you Guillaume!

**Comment by joshuahansel** [comment, marked as answer]
I don't have your input file or stacktrace, but I think the error you're probably hitting is in `TensorMechanicsAction`. If you want to check the suspicion you have, you can check whether the task `THM:build_mesh` appears before `validate_coordinate_systems` when you do `Debug/show_actions=true`.

**Reply by jcstonehillAMA** [reply]
Excellent! Thanks for the advice. I'll check this out and report back. I'm a fairly novice MOOSE user and didn't know THM was meant to couple through Multiapps. I'll read up more on Multiapps for the future.

**Reply by joshuahansel** [reply]
What physical problem are you working toward? You're trying to couple tensor mechanics to a THM heat structure. Does that mean you want to couple to thermal hydraulics? And what kind of coupling between physical domains are you trying to achieve?

**Reply by jcstonehillAMA** [reply]
Joshua, I apologize for the late reply. I've been on leave.

As I mentioned, I'm new to MOOSE. The goal was to create a model of a NERVA-derived fuel element. The solid domains would have volumetric heating and heat conduction. This solid domain was setup as a heat structure from an exodus file. Then I used heat transfer components to transfer the heat to multiple FlowChannel1Phase components (one for each of the flow channels in the physical fuel element). 

I was able to accomplish this. The next step was to see if I could use the results to determine stresses/strains from the thermal expansion combined with an applied load, which is when I ran into problems. I've been learning the MOOSE suite by simply examining examples and trying to mimic them to fit my problem. I had not considered the multi-app functionality and I'm curious if that would solve my issues! When I get a chance, I'll give it an attempt.
